crates/ruma-api-macros/src/request/outgoing.rs

```diff
@@ -15,39 +15,33 @@ impl Request {
 
         let method = &self.method;
         let error_ty = &self.error_ty;
-        let request_path_string = if self.has_path_fields() {
-            let mut format_string = self.path.value();
-            let mut format_args = Vec::new();
 
-            while let Some(start_of_segment) = format_string.find(':') {
-                // ':' should only ever appear at the start of a segment
-                assert_eq!(&format_string[start_of_segment - 1..start_of_segment], "/");
+        let (unstable_path, r0_path, stable_path) = if self.has_path_fields() {
+            let path_format_args_call_with_percent_encoding = |s: &LitStr| -> TokenStream {
+                let (s, a) = util::path_format_args_call(s.value(), &percent_encoding);
 
-                let end_of_segment = match format_string[start_of_segment..].find('/') {
-                    Some(rel_pos) => start_of_segment + rel_pos,
-                    None => format_string.len(),
-                };
-
-                let path_var = Ident::new(
-                    &format_string[start_of_segment + 1..end_of_segment],
-                    Span::call_site(),
-                );
-                format_args.push(quote! {
-                    #percent_encoding::utf8_percent_encode(
-                        &::std::string::ToString::to_string(&self.#path_var),
-                        #percent_encoding::NON_ALPHANUMERIC,
-                    )
-                });
-                format_string.replace_range(start_of_segment..end_of_segment, "{}");
-            }
+                quote! {
+                    format_args!(#s, #(#a),*)
+                }
```

Comment:
What I meant in that other review comment is that this `quote!` call could be moved into `util::path_format_args_call` so it has a simpler signature. The closure will still be needed, but will be shorter.